### PR TITLE
amp-bind: scanAndApply() without DOM walking

### PR DESCRIPTION
**Pull request by choumx**
Alternative to #17203.

Under a new experiment 'faster-bind-scan':
- Add `i-amphtml-binding` attribute to elements with bindings i.e. `[foo]` or `data-amp-bind-foo`
- Use `querySelectorAll('[i-amphtml-binding]')` instead of TreeWalker
- Move removal of old bindings out of critical path of `scanAndApply` as another optimization

Also simplifies some of the logic for DOM scanning in bind-impl.js.

### CODE REVIEW

**Comment by jridgewell** [comment]
> Add i-amphtml-binding attribute to elements with bracket attributes [foo]

`data-bind-foo`, too?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -360,21 +356,63 @@ export class Bind {
...
+      const added = bindings.length;
+      if (added === 0) {
+        return cleanup(0);
+      }
+      promise = this.sendBindingsToWorker_(bindings).then(() => {
+        return this.evaluate_()
+            .then(results => this.applyElements_(results, addedElements));
+      }).then(() => {
+        // Remove bindings at the end to reduce evaluation/apply latency.
+        // Don't chain this promise since this is a non-blocking clean up task.
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: This doesn't really matter, since it's just a `Promise.resolve()`.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -722,18 +738,50 @@ export class Bind {
...
+   * @param {!Array<!BindBindingDef>} outBindings
+   * @return {boolean} Returns true if `element` contains more than `quota`
+   *     bindings. Otherwise, returns false.
+   */
+  scanElement_(element, quota, outBindings) {
+    let quotaExceeded = false;
+    let boundProperties = this.boundPropertiesInElement_(element);
+    // Stop scanning once |limit| bindings are reached.
+    if (boundProperties.length > quota) {
+      boundProperties = boundProperties.slice(0, quota);
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Nit: we could turn `boundProperties` into a constant if we change this to `boundProperties.length = quota`

File: extensions/amp-list/0.1/amp-list.js [hunk 3]
```diff
@@ -337,17 +335,22 @@ export class AmpList extends AMP.BaseElement {
...
    * @private
    */
-  updateBindingsForElements_(elements) {
+  updateBindings_(elements) {
     const binding = this.element.getAttribute('binding');
     // "no": Always skip binding update.
     if (binding === 'no') {
       return Promise.resolve(elements);
     }
+    const updateWith = bind => {
```
Comment (jridgewell) [inline extensions/amp-list/0.1/amp-list.js]: Why was this moved to a closure?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 1 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: This is an old comment, thanks.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 2 above)
Comment (choumx) [reply extensions/amp-bind/0.1/bind-impl.js]: Good idea.

File: extensions/amp-list/0.1/amp-list.js (hunk 3 above)
Comment (choumx) [reply extensions/amp-list/0.1/amp-list.js]: Just for readability/conciseness.
